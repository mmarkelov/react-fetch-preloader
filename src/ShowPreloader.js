/* global window */
import React from 'react';
import PropTypes from 'prop-types';

const INTERVAL = 50;

class ShowPreloader extends React.Component {
  static propTypes = {
    fetch: PropTypes.func,
    url: PropTypes.string.isRequired,
    duration: PropTypes.number,
    loaded: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired,
    errored: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    preloader: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  };

  static defaultProps = {
    fetch: typeof window !== 'undefined' ? window.fetch : () => {},
    duration: 500,
    errored: () => <span>ERROR</span>,
    preloader: () => <span>Loading...</span>,
  };

  state = {
    responseTime: 0,
    isLoading: false,
    showPreloader: false,
    data: null,
    error: null,
  };

  componentDidMount() {
    const { url, loaded } = this.props;
    if (!url) {
      throw new Error('url param is required!');
    }
    if (!loaded) {
      throw new Error('loaded param is required!');
    }
    this.makeResponse(url);
    this.intervalId = setInterval(this.checkResponse, INTERVAL);
  }

  componentDidUpdate(prevProps) {
    const { url } = this.props;
    if (url !== prevProps.url) {
      this.makeResponse(url);
      this.intervalId = setInterval(this.checkResponse, INTERVAL);
    }
  }

  componentWillUnmount() {
    clearInterval(this.intervalId);
  }

  checkResponse = () => {
    const { duration } = this.props;
    const { responseTime, isLoading, showPreloader } = this.state;
    this.setState(
      prevState => ({
        responseTime: prevState.responseTime + INTERVAL,
      }),
      () => {
        if (responseTime >= duration && isLoading) {
          if (!showPreloader) {
            this.setState({ showPreloader: true });
          }
          clearInterval(this.intervalId);
        }
      },
    );
  };

  makeResponse = url => {
    const { fetch } = this.props;
    this.setState({ responseTime: 0, isLoading: true });
    fetch(url)
      .then(resp => resp.json())
      .then(data => {
        this.setState({ isLoading: false, data, showPreloader: false });
        clearInterval(this.intervalId);
      })
      .catch(error => {
        this.setState({ isLoading: false, error, showPreloader: false });
        clearInterval(this.intervalId);
      });
  };

  renderContent = () => {
    const { loaded, errored } = this.props;
    const Loaded = loaded;
    const Error = errored;
    const { data, isLoading, error } = this.state;
    if (error) {
      return <Error error={error} />;
    }
    return !isLoading && data ? <Loaded data={data} /> : null;
  };

  render() {
    const { showPreloader } = this.state;
    const { preloader } = this.props;
    const Preloader = preloader;

    return showPreloader ? <Preloader /> : this.renderContent();
  }
}

export default ShowPreloader;
