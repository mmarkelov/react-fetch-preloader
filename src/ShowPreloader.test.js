import React from 'react';
import { shallow } from 'enzyme';
import ShowPreloader from './ShowPreloader';

const Loaded = () => <div>Loaded</div>;
const Error = () => <div>Error</div>;
const URL = 'https://randomuser.me/api/?results=10';

const fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    ok: true,
    json: () => ({
      data: 'Some data',
    }),
  }),
);

const rejectedFetch = jest.fn().mockImplementation(() =>
  Promise.reject({
    error: true,
    json: () => ({
      data: 'Error',
    }),
  }),
);

describe('ShowPreloader', () => {
  it('should throw Error without url prop', () => {
    try {
      shallow(<ShowPreloader loaded={Loaded} />);
    } catch (e) {
      expect(e.message).toEqual('url param is required!');
    }
  });

  it('should throw Error without loaded prop', () => {
    try {
      shallow(<ShowPreloader url={URL} />);
    } catch (e) {
      expect(e.message).toEqual('loaded param is required!');
    }
  });

  it('makeResponse should be called when url was changed', () => {
    let props = { url: 'prev_url', loaded: Loaded, fetch };
    const component = shallow(<ShowPreloader {...props} />);
    const makeResponse = jest.spyOn(component.instance(), 'makeResponse');
    component.setProps({ ...props, url: 'new_url' });
    expect(makeResponse).toHaveBeenCalledWith('new_url');
  });

  it('should render preloader component while fetching', async () => {
    const component = shallow(
      <ShowPreloader url={URL} loaded={Loaded} fetch={fetch} duration={0} />,
    );

    component.instance().checkResponse();
    expect(component.html()).toEqual('<span>Loading...</span>');
  });

  it('should render loaded component if responsed finished', async () => {
    const component = await shallow(
      <ShowPreloader url={URL} loaded={Loaded} fetch={fetch} />,
    );
    await component.instance().makeResponse();
    expect(component.html()).toBe('<div>Loaded</div>');
  });

  it('should render error component if response failed', async () => {
    const component = await shallow(
      <ShowPreloader
        url={URL}
        loaded={Loaded}
        fetch={rejectedFetch}
        errored={Error}
      />,
    );
    await component.instance().makeResponse();
    process.nextTick(() => {
      expect(component.html()).toBe('<div>Error</div>');
    });
  });

  it('should render correctly with all props', () => {
    const component = shallow(
      <ShowPreloader url={URL} loaded={Loaded} fetch={fetch} />,
    );
    expect(component).toMatchSnapshot();
  });
});

afterAll(() => setTimeout(() => process.exit(), 1000));
