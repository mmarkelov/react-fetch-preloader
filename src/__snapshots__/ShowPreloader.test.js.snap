// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowPreloader should render correctly with all props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowPreloader
    duration={500}
    errored={[Function]}
    fetch={
      [MockFunction] {
        "calls": Array [
          Array [
            "prev_url",
          ],
          Array [
            "new_url",
          ],
          Array [
            "https://randomuser.me/api/?results=10",
          ],
          Array [
            "https://randomuser.me/api/?results=10",
          ],
          Array [
            undefined,
          ],
          Array [
            "https://randomuser.me/api/?results=10",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {},
          },
          Object {
            "isThrow": false,
            "value": Promise {},
          },
          Object {
            "isThrow": false,
            "value": Promise {},
          },
          Object {
            "isThrow": false,
            "value": Promise {},
          },
          Object {
            "isThrow": false,
            "value": Promise {},
          },
          Object {
            "isThrow": false,
            "value": Promise {},
          },
        ],
      }
    }
    loaded={[Function]}
    preloader={[Function]}
    url="https://randomuser.me/api/?results=10"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
